[Processing] GRASS7 complete i.* modules (implements Redmine #5722)

medspx opened the pull request:
# About

Like PR #2904, here is the implementation of the last GRASS7 i.\* (imagery) modules for QGIS Processing. For the list of algorithms, just read commits comments...

This should implement [Redmine #5722 feature request](https://hub.qgis.org/issues/5722).

I would like to have it (if there are not too much bugs) merged into QGIS master before the feature freeze (in 15 days) in order to be able to fix bugs on it and to be able to say that now QGIS Processing implements every GRASS7 algorithm !

It was really hard to implement because those algorithms needs lots of data (and because I am an imagery newbie). I've tried to use real world data (ASTER/LANDSAT/MODIS/SPOT) in order to avoid insane results.

PS: I will be disconnected from the Internet for a week so don't expect code changes/answers before Sun 14-05-2016 ;-)
# Code

I have made more factorization effort in this PR (compared to GRASS7 r.\* and v.\* algorithms). The main code is in ext/i.py and each algorithm which needs a dedicated logic generally uses code from i.py. I could relocate some code into Grass7Utils but I have chosen to stick it to i.py for the moment because I intend to get further with factorization for v.\* and r.\* algorithms once all the unit tests will be implemented. It will be easier to spot bugs resulting from factorization...
# Tests

This time, I have taken time to make unit tests using the new Processing tests suite. I have created about half of the tests (25 tests for about 45 algorithms). I have also made some minor modifications to the test suite in order to implement a maximum of tests for those modules:
- fix raster support ('gdal' instead of 'ogr' for opening a QgsRasterLayer).
- support external files.
- improve graphical test generator with external files and multiple input datatype detection.

I know that I could/should have done a dedicated PR for this but I have not found the time to do it properly (and I would like to include tests into this PR). If you request a dedicated PR for test suite improvements, I will be able to work on it in about a week...

For data, I have generated 4 random small TIFF rasters (about 100x100 pixels) and I have tried to use them everywhere. Some i.\* algorithms needs a lot of different rasters layers (at least 6 or 7 for Landsat algorithms). For those ones, I haven't already found the good dataset to use in order to a get a better result than a NULL raster.

To get further, we also need a support for OutputDirs in the test suite. i.\* modules often generate lots of raster files in a directory. For the moment, there is no way to predict raster names . So all algorithms which have an OutputDir output are not tested.

You can read a full recap about unit tests in the TODO file.
# What has been excluded ?
- i.ortho.*: I could not find a way to use this. GRASS7 Documentation seems to be outdated (404 errors for some links to other i.ortho algorithms). I think it is not possible to use them in GRASS7.
- i.spectral: it needs absolutely a graphical window within GRASS7 to work.
- i.target: it is an internal algorithm used with i.rectify. The Processing i.rectify implementation use it internally.
# What is broken ?
- i.cca seems to be broken (or I have just not find the good dataset for it).
- i.aster.toar seems to be broken in GRASS7 because it always reports that you need 15 rasters to use it, even when you have a true ASTER L1B hdf file with 15 layers. I have tried to use 16 rasters but it doesn't work too. I will report a bug in the GRASS bugtracker.
- For some algorithms, there is a Python bug when you try to launch them multiple times from the same dialog. I haven't found from where it comes even if I suspect that there is a dynamic (virtual) parameter which has not been correctly reset.
- I have not made tests with models so there could be bugs (at least for the previous problem).
# What's next ?

Well, after this PR, we can say that QGIS Processing implements all of the GRASS7 algorithms that are implementable in QGIS. Once done, we need a full cover for unit tests. It should be fastidious to make but not so complicated thanks to the graphical unit test generator. I was able to write 25 tests in a day of work.

Once there will be enough unit tests, I will work on code factorization. It should be easier to work on it with regression detection (thanks to unit tests). Now that r._, v._ and i.\* algorithms are implemented, code factorization should be more complete.

Once code factorization will be done, I think that I will work on the last GRASS7 bugs and the Processing GRASS7 implementation should be done !

Commit:
Add i.albedo and i.cluster algorithms

Commit:
Add i.colors.enhance algorithm

Commit:
Add i.eb.evapfr algorithm

Commit:
Add i.eb.eta algorithm

Commit:
Add i.eb.hsebal01 algorithm

Commit:
Add i.eb.netrad algorithm

Commit:
Add i.eb.soilheatflux algorithm

Commit:
Add i.biomass algorithm

Commit:
Add i.emissivity algorithm

Commit:
Add i.evapo.mh algorithm

Commit:
Add i.evapo.pm algorithm

Commit:
Add i.evapo.pt algorithm

Commit:
Add i.evapo.time algorithm

Commit:
Add i.gensig algorithm

Commit:
Add i.gensigset algorithm

Commit:
Add i.group algorithm

Commit:
Add i.evapo.mh parameters detection

Commit:
Add i.image.mosaic (needs Redmine #5742 to be implemented)

Commit:
Add i.landsat.toar algorithm

Commit:
Add i.vi algorithm

Commit:
Add i.tasscap algorithm

Commit:
Add i.smap algorithm

Commit:
Add i.segment algorithm

Commit:
Add i.pca algorithm

Commit:
Add i.topo.corr algorithm

Commit:
Add i.pansharpen algorithm

Commit:
Add i.oif algorithm

Commit:
Add i.cca algorithm (segfaults from GRASS)

Commit:
Add i.maxlik algorithm

Commit:
Add i.modis.qc algorithm

Commit:
Add i.in.spotvgt algorithm

Commit:
Add i.landsat.acc algorithm

Commit:
Add i.rectify algorithm

Commit:
Add i.aster.toar algorithm (need to fix a GRASS7 bug about number of files)

Commit:
Fix some algorithms

Commit:
Add first set of unit tests

Commit:
Improve tests: add external file support into AlgorithmsTest

Commit:
Expand external file support to graphical unit test generator

Commit:
Fix multiple input parameter type detection for graphical unit test generator

Commit:
Add last set of tests

Commit:
Fix i.atcor algorithm

Commit:
Update TODO file

pcav commented:
Thanks a lot @medspx !

pcav commented:
@volaya @alexbruy

m-kuhn commented:
@medspx It looks like you had a thorough look into the processing test suite. You are probably the first one who looked into it in that much detail.
If you have any feedback, I am very happy to hear it.

volaya commented:
Awesome work. Looks good to me. +1 to merge

There seems to be an error in Travis related to the tests. Maybe @m-kuhn can comment on that before merging

m-kuhn commented:
Absolutely, brilliant stuff.
@medspx seems to be quite knowledgable of the testsuite already, I'm waiting for your call if you need assistance to figure out details about it.

medspx commented:
Hello, I'm back !

@m-kuhn I don't know why there is the exec error in Travis ?
On my computer, `make test` build all the tests without errors (at least on Processing suite). Is there really a problem with the exec statement ? Perhaps should we remove the first tuple ?
`exec(('\n'.join(defs['expectedFailure'][:-1])), globals(), locals())`
to
`exec('\n'.join(defs['expectedFailure'][:-1]), globals(), locals())` but I can't find any difference when I launch `make test`.

Should I force another build ?

Commit:
Try this exec modification

Commit:
Try without subfunction

medspx commented:
Hello,

I've tried to get rid of the exec problem in the latest build... But I am facing problems with test suite: it seems that some algs are just crashing and for the majority of the others, the output rasters have a different hash.

Should I remove the tests in order to have the algs implemented into next release (deadline is tomorrow) ? If so, I could try to add it later (during the bug fixing phase).

What do you think ? @m-kuhn 

Otherwise, I could try to reproduce the test failures found in travis on my computer (but don't know how to start)...

Commit:
Fix i.at.corr test

Commit:
Add ugly debug prints to try to find what is travis-ci problem...

Commit:
Temporary make Grass7Utils more verbose for more travis debug

medspx commented:
Hello,

I've taken time to investigate on the travis side (by recreating the whole environment in a docker container) and here is what I have found so far:
- to deal with the initial problem of exec, I have just removed the nested function (and repeated code which is bad but I haven't found why Python throw an error on this side).
- travis-ci uses GRASS 7.0.2 (thanks to osgeo4travis) which is downloaded in the pre-build process. The GRASS version seems to be a problem for rasterhashes because raster metadata depends on GRASS version: there is a metadata which just reports the GRASS version. So if you use a different GRASS version from osgeo4travis (which is the case for me, developping on Debian testing with GRASS 7.0.4), you will inevitability get different rasterhashes and the unit tests will fail. I have compared the raster produced by each version of GRASS (Debian vs osgeo4travis) and the data are the same, only GRASS version metadata is different. To deal with this, you will need to use osgeo4travis GRASS.  I will take time to update test documentation for this side and to produce a new set of tests.
- furthermore, some GRASS modules needs libraries that are not (yet) available in travis environment (read [this build log](https://travis-ci.org/qgis/QGIS/jobs/132592980#L1917)). For the moment, i.zc and i.fft (and others) needs `libfftw3-3` package, otherwise, those GRASS modules will fail and the unit tests will also fail. I've made further tests and only this library is required for the tests described in this PR. @m-khun, is it possible to add this package (or a true libfftw3.so.3 somewhere on the LD_LIBRARY_PATH) in osgeo4travis or in the travis-ci build environment ?

Ok, now that I have a complete travis-ci environment, I will update this PR with the fixes...

Commit:
Fix rasterhashes value with osgeo4travis values

m-kuhn commented:
Thanks @medspx for all the info!

> to deal with the initial problem of exec, I have just removed the nested function (and repeated code which is bad but I haven't found why Python throw an error on this side).

Is there a particular commit you can link to?

> travis-ci uses GRASS 7.0.2 (thanks to osgeo4travis) which is downloaded in the pre-build process. The GRASS version seems to be a problem for rasterhashes because raster metadata depends on GRASS version: there is a metadata which just reports the GRASS version. So if you use a different GRASS version from osgeo4travis (which is the case for me, developping on Debian testing with GRASS 7.0.4), you will inevitability get different rasterhashes and the unit tests will fail. I have compared the raster produced by each version of GRASS (Debian vs osgeo4travis) and the data are the same, only GRASS version metadata is different. To deal with this, you will need to use osgeo4travis GRASS. I will take time to update test documentation for this side and to produce a new set of tests.

Interesting. We probably want to exclude the metadata version from the rasterhash then, what do you think?

> furthermore, some GRASS modules needs libraries that are not (yet) available in travis environment (read this build log). For the moment, i.zc and i.fft (and others) needs libfftw3-3 package, otherwise, those GRASS modules will fail and the unit tests will also fail. I've made further tests and only this library is required for the tests described in this PR. @m-kuhn, is it possible to add this package (or a true libfftw3.so.3 somewhere on the LD_LIBRARY_PATH) in osgeo4travis or in the travis-ci build environment ?

Have a look at .travis.yml in the toplevel directory and ci/travis/linux/qt5/before_install.sh for pip install and apt install commands.

Commit:
Add libfftw3-3 to travis-ci environment (fingercross: should pass)

Commit:
Fix exec statement for AlgorithmsTestBase.py
On travis-ci environment, Python version seems to be affected by [this bug](https://bugs.python.org/issue21591).
One way to fix it is to use the old statement instead of exec() function.

Commit:
fix .travis config indentation

medspx commented:
Hello,

Ok qt4 branch is building and pass all tests. I've tried to fix the exec statement (see in [829807f](https://github.com/qgis/QGIS/pull/3062/commits/829807fc2bbbed8fdf35039d2e56f99c11873ef8)) but it is not valid for python3.

Unfortunately, for raster hashes, we do not have other choice than using the osgeo4travis version of GRASS because r.out.gdal mandatory includes a "Generated with" Metadata (no option to exclude it [as you can read in the code](https://trac.osgeo.org/grass/browser/grass/branches/releasebranch_7_0/raster/r.out.gdal/export_band.c#L241).

Travis still tells that the build failed. Is the qt5 build mandatory ? If this is the case, I'll have to find a way to find a solution for both python versions (perhaps using what I have done [in this commit](https://github.com/qgis/QGIS/pull/3062/commits/c3ca1bfd4dc82309f9d833f1c04c4d7a57584515) which is not compact but should work. Another solution would be to upgrade the Python 2.7 version to a more recent one ? @m-kuhn @volaya @alexbruy What do you think ?

Commit:
Fix qt5 travis build and update README for Processing tests

m-kuhn commented:
> Ok qt4 branch is building and pass all tests. I've tried to fix the exec statement (see in 829807f) but it is not valid for python3.

What exactly is the problem there, I think this is already running now for the python 3 travis builds for other algorithm providers. Why is this different for grass?

> Unfortunately, for raster hashes, we do not have other choice than using the osgeo4travis version of GRASS because r.out.gdal mandatory includes a "Generated with" Metadata (no option to exclude it as you can read in the code).

This will limit us to ever upgrade grass, I think we'll have to find another solution that calculates a hash without taking the metadata into account. Or find some other good logic.

medspx commented:
Ok, 

@m-kuhn I have taken time to investigate on the raster hashes side. It was not a problem with metadata (also they were different) since we use gdal ReadAsArray method from the opened dataset. The method returns a numpy array which is completely independent from metadata (just raster values as an array). I made comparisons between two rasters produced with osgeo4travis GRASS and Debian testing GRASS 7.0.4 and all the values were identical in the resulting arrays. The only difference was on the NaN value: even if you have NaN values at the same place in the array, the function numpy.array_equal returned False. Converting NaN values to number (with numpy.nan_to_num) fixed the problem. Perhaps osgeo4travis GRASS r.out.gdal binary produces rasters with an internal definition of NaN values which is different from Debian GRASS (that's what I am supposing), resulting in a data buffer structure with differences leading to different hashes. 

The nan_to_num conversion just fix the problem. I have made tests under MS-Windows (with OsGeo4W64) and the raster hashes are equals than those produced under GNU/Linux. I will upload a fix soon (once I will have remade all the tests).

Commit:
Remove NaN values from raster before hash calculation

Commit:
Exec things...

Commit:
Blacklist travis-ci qt5 tests and remove i.fft test

medspx commented:
Hello,

I think it is ready to merge now !

 Here is what I have finally done:
* on qt4 build, only the i.fft test produces a different raster hash in travis than the one produced with the nan_to_num fix. I suppose it is linked to the version of libfftw which is quite old in travis environment. Anyway, I have commented the test for future re-use, once we have further details about what's wrong here.
* for the qt5 build I have no other choice than blacklisting GRASS processing test. Actually,  GRASS Processing is not python3 ready and this problem [triggers errors in travis build](https://travis-ci.org/qgis/QGIS/jobs/133585291#L798). I don't have the time to study the python3 port of this part of Processing before feature freeze (and I really want to have i.* modules in QGIS 2.16). Furthermore, I want to work on PR #2882 before working on the python3 port because I believe that it will be more difficult to fix on python3 (let's find a solution for python2 and it will be easier to port, there is a lot of encoding problem in this PR).

So here we are:
* We have all of the i.* GRASS7 modules.
* We have an embryo of Processing GRASS7 test suite which should be stable and which can be improved in the future.
* Once PR #2882 will be fixed (or closed because too much work/no solution), GRASS7 Processing should be "ported" to python3.
* Once the port is done, there should be no problem to de-blacklist the tests in qt5 build.

@m-kuhn @volaya @alexbruy 
Anybody to merge this ?

m-kuhn commented on python/plugins/processing/algs/grass7/TODO:
```diff
@@ -1,9 +1,70 @@
-* http://trac.osgeo.org/grass/wiki/Grass7/NewFeatures#Renamedmodules
+TODO List for GRASS7 algorithms support into QGIS Processing
+
+Unit tests
```
Just a small thing, if you rename this file to TODO.md it will be formatted automatically, especially since the syntax seems to be markdown already.

m-kuhn commented:
Great stuff, thank you so much!

nyalldawson commented:
@medspx this is failing on the OSX build - see http://dash.orfeo-toolbox.org/testDetails.php?test=41599322&build=230843
